mlkem.go

```diff
@@ -0,0 +1,374 @@
+
+	// encapsulationKeySizeMLKEM768 is the size of an ML-KEM-768 encapsulation key.
+	encapsulationKeySizeMLKEM768 = 1184
+
+	// ciphertextSizeMLKEM1024 is the size of a ciphertext produced by ML-KEM-1024.
+	ciphertextSizeMLKEM1024 = 1568
+
+	// encapsulationKeySizeMLKEM1024 is the size of an ML-KEM-1024 encapsulation key.
+	encapsulationKeySizeMLKEM1024 = 1568
+)
+
+// SupportsMLKEM768 returns true if ML-KEM-768 is supported on this platform.
+func SupportsMLKEM768() bool {
+	if vMajor >= 3 && vMinor >= 5 {
+		return supportsMLKEM768()
+	}
+	return false
+}
+
+// SupportsMLKEM1024 returns true if ML-KEM-1024 is supported on this platform.
+func SupportsMLKEM1024() bool {
+	if vMajor >= 3 && vMinor >= 5 {
+		return supportsMLKEM1024()
+	}
+	return false
+}
+
+var supportsMLKEM768 = sync.OnceValue(func() bool {
+	sig, _ := ossl.EVP_KEYMGMT_fetch(nil, _KeyTypeMLKEM768.ptr(), nil)
+	if sig != nil {
+		ossl.EVP_KEYMGMT_free(sig)
+		return true
+	}
+	return false
+})
+
+var supportsMLKEM1024 = sync.OnceValue(func() bool {
+	sig, _ := ossl.EVP_KEYMGMT_fetch(nil, _KeyTypeMLKEM1024.ptr(), nil)
+	if sig != nil {
+		ossl.EVP_KEYMGMT_free(sig)
+		return true
+	}
+	return false
+})
+
+// DecapsulationKeyMLKEM768 is the secret key used to decapsulate a shared key
+// from a ciphertext. It includes various precomputed values.
+type DecapsulationKeyMLKEM768 [seedSizeMLKEM]byte
+
+// GenerateKeyMLKEM768 generates a new decapsulation key, drawing random bytes from
+// the default crypto/rand source. The decapsulation key must be kept secret.
+func GenerateKeyMLKEM768() (DecapsulationKeyMLKEM768, error) {
+	var dk DecapsulationKeyMLKEM768
+	_, err := rand.Read(dk[:])
+	if err != nil {
+		return DecapsulationKeyMLKEM768{}, err
+	}
+	return dk, nil
+}
+
+// NewDecapsulationKeyMLKEM768 expands a decapsulation key from a 64-byte seed in the
+// "d || z" form. The seed must be uniformly random.
+func NewDecapsulationKeyMLKEM768(seed []byte) (DecapsulationKeyMLKEM768, error) {
+	if len(seed) != seedSizeMLKEM {
+		return DecapsulationKeyMLKEM768{}, errors.New("mlkem: invalid seed size")
+	}
+
+	var dk DecapsulationKeyMLKEM768
+	copy(dk[:], seed)
+	return dk, nil
+}
+
+// Bytes returns the decapsulation key as a 64-byte seed in the "d || z" form.
+//
+// The decapsulation key must be kept secret.
+func (dk DecapsulationKeyMLKEM768) Bytes() []byte {
+	return dk[:]
+}
+
+// Decapsulate generates a shared key from a ciphertext and a decapsulation
+// key. If the ciphertext is not valid, Decapsulate returns an error.
+//
+// The shared key must be kept secret.
+func (dk DecapsulationKeyMLKEM768) Decapsulate(ciphertext []byte) (sharedKey []byte, err error) {
+	return performDecapsulation(ossl.NID_ML_KEM_768, dk[:], ciphertext)
+}
+
+// EncapsulationKey returns the public encapsulation key necessary to produce
+// ciphertexts.
+func (dk DecapsulationKeyMLKEM768) EncapsulationKey() EncapsulationKeyMLKEM768 {
+	pkey, err := createMLKEMPrivateKey(ossl.NID_ML_KEM_768, dk[:])
```

Comment:
This function is almost the same as `DecapsulationKeyMLKEM768.EncapsulationKey`, factor it out.